Add robot mode publisher, masterboard publisher and waitForURProgram example function

felixvd opened the pull request:
This is a port of https://github.com/Zagitta/ur_modern_driver/pull/27 and includes #181. I am reopening it here as proposed by @gavanderhoorn in the original PR. I tested it on our robots, we had it running in production, and this rebase still compiles. 

In the meantime, @ipa-nhg suggested to put `RobotModeDataMsg` in ur_msgs, which I agree with, but I only saw the comment after rebasing this. Opinions?

gavanderhoorn commented:
Hi @felixvd, thanks for the PR and thanks for porting it.

CI seems to fail because of Clang formatting. Could you add a commit that fixes the formatting?

Configuration is available [here](https://github.com/ros-industrial/ur_modern_driver/blob/e2d81cf2b88237701931e02dbf05780d135c81c4/.clang-format).

gavanderhoorn commented:
> In the meantime, @ipa-nhg suggested to put `RobotModeDataMsg` in ur_msgs, which I agree with, but I only saw the comment after rebasing this. Opinions?

I would definitely be in favour of this.

All msg defs for the UR driver(s) should go into `ur_msgs` for now. Introducing messages specifically in this package should be avoided.

The PR becomes slightly more complicated (as CI will always fail without getting `ros-industrial/universal_robot` from source as well), but we can probably handle that.

felixvd commented:
As you said, this CI will fail, but here you go. It compiled on my machine.

felixvd commented:
With https://github.com/ros-industrial/universal_robot/pull/395 merged, should this Travis build succeed? Does anything need to be triggered manually?

gavanderhoorn commented:
I've restarted the build.

miguelprada commented:
I'd like to suggest merging this without including the `wait_for_program.h` convenience functions. Its obvious that they can be helpful to some, but it feels kind of out of scope for me. What do you think @gavanderhoorn, @ipa-nhg?

Regarding the failing checks, I opened #266 to discuss that problem.

felixvd commented:
For the record, I put it in because a) the feature of the [original PR](https://github.com/Zagitta/ur_modern_driver/pull/27) seemed incomplete without the convenience function, and b) it is relevant for anyone who wants to use the /URscript topic. 

I also think tiny examples serve essentially the same purpose as a Readme, but that's a question of style.

miguelprada commented:
> For the record, I put it in because 1) the feature of the [original PR](https://github.com/Zagitta/ur_modern_driver/pull/27) seemed incomplete without the convenience function, and 2) it is relevant for anyone who wants to use the /URscript topic.

Having the information published is definitely relevant, but I'm not sold on providing a convenience function which basically waits for a message and checks a boolean field. I think that's considered as something any experienced ROS user will implement themselves in their nodes (e.g. no convenience `get_joint_state()` function is provided).

> I also think tiny examples serve essentially the same purpose as a Readme, but that's a question of style.

Indeed. I would prefer pointing out which field in the message should be checked in the Readme, but others may disagree. Let's wait to hear from the other maintainers.

felixvd commented:
Totally reasonable. My personal experience is that I appreciate easy-to-find convenience functions because I would waste time reimplementing them :)

gavanderhoorn commented:
@felixvd: late, but: with the merge of #269, a rebase of your PR *should* update our CI configuration such that your PR should receive a green checkmark from Travis.

felixvd commented:
I rebased it and the checks did pass. Now to decide if you want to keep the `wait_for_program.h` convenience header.

miguelprada commented on src/ros/mb_publisher.cpp:
```diff
@@ -16,12 +16,30 @@ void MBPublisher::publish(ur_msgs::IOStates& io_msg, SharedMasterBoardData& data
 {
   // note that this is true as soon as the drives are powered,
-  // even if the breakes are still closed
+  // even if the breaks are still closed
```
brakes

felixvd replied:
:)

miguelprada commented on README.md:
```diff
@@ -168,13 +169,9 @@ Should be compatible with all robots and control boxes with the newest firmware.
-* Simulated UR5 running 1.6.08725
+*Real UR5 and UR3 with 3.5.1.1.10661 (Dec 13 2017)
+
+*Real UR10 with CB2 running 1.8.14035
```
Where are these changes coming from? The new version with just two lines doesn't match the previous, more detailed, version.

felixvd replied:
I don't know why the lines were deleted either; maybe something went wrong with the rebase. I only remember adding line 172. I updated this and included our newer robot.

miguelprada replied:
Probably rebasing shenanigans, yes. :+1:

miguelprada commented on README.md:
```diff
@@ -168,6 +169,8 @@ Should be compatible with all robots and control boxes with the newest firmware.
 ### Tested with:
 
+* Real UR5 with 3.9.0.46176 (Mar 01 2019)
+* Real UR5 and UR3 with 3.5.1.1.10661 (Dec 13 2017)
```
I think you have an extra 1 here, should be `3.5.1.10661`.

miguelprada commented on src/ros/mb_publisher.cpp:
```diff
@@ -16,12 +16,30 @@ void MBPublisher::publish(ur_msgs::IOStates& io_msg, SharedMasterBoardData& data
+  // --- Publishes the data in robot-specific format.
+  ur_msgs::MasterboardDataMsg msg;
+  msg.digital_input_bits = 0;   // TODO: Set this.
+  msg.digital_output_bits = 0;  // TODO: Set this.
```
There's also this, which is not straightforward to fix and not sure it's worth the effort.

Not straightforward because properly setting this would require changing the `MasterboardDataMsg` to accomodate 18 bits in these fields.

Not worth the effort because this information is already published in the `IOStates` messages.

gavanderhoorn replied:
I'm confused: aren't the bits just part of a bitmask in the original data (according to the protocol spec these fields are 4 byte integers)? If so: these fields should just be set to the value in the original message.

@miguelprada: are you suggesting that each individual pin should be addressable in `digital_input_bits` (ie: make it a list or something like that)?

miguelprada replied:
To be honest, I haven't looked much into this. Even if the protocol marks them as integers, I reckon that these are parsed into bitmasks of size [10](https://github.com/ros-industrial/ur_modern_driver/blob/093e1fbd2a6f81286f318aad20f539b5eb494259/include/ur_modern_driver/ur/master_board.h#L63-L64) or [18](https://github.com/ros-industrial/ur_modern_driver/blob/093e1fbd2a6f81286f318aad20f539b5eb494259/include/ur_modern_driver/ur/master_board.h#L84-L85) for a reason.

Those 18 bits not fitting in the `MasteRboarDataMsg` was why I said it wasn't straightforward to fix. I saw you are already trying to fix it, though :wink:

miguelprada commented:
I'll leave it to @gavanderhoorn or @ipa-nhg to merge as is or comment what they think about the convenience functions header.

ipa-nhg commented:
I prefer to merge this without the  `wait_for_program.h` header, it is a very useful tool but from my point of view should not be part of the driver.
However it is not introducing extra dependencies to the package, then it is just a preference because of the content, if @miguelprada you wanted to keep it is also ok for me.

gavanderhoorn commented:
With two out of three maintainers requesting to not merge the utility functions into the driver, I think we should leave them out @felixvd.

What we can do is introduce a `ur_modern_driver_utilities` package (or something similarly named) and host them there. Would that work for you?

felixvd commented:
I fixed the version number typo and removed the `wait_for_program.h` convenience header and moved it into [this gist](https://gist.github.com/felixvd/4d5db1c297a3435cabb40c857eb6ead5) instead. I think a separate package might be a bit much for this little file, but feel free to reuse it in any way you like!

Cheers

gavanderhoorn commented:
> I think a separate package might be a bit much for this little file, but feel free to reuse it in any way you like!

For this single one: yes. But if there's a package, it might invite others to contribute other convenience functions, tools, etc.

gavanderhoorn commented:
Let's wait on Travis and merge.

gavanderhoorn commented:
Travis is happy, I'm happy, so merging.

Note: squash merging as most commits are fixups.

gavanderhoorn commented:
Thanks @felixvd. And many thanks for iterating.

:+1:

miguelprada commented:
Yes, thanks for your patience @felixvd!
